Create export script

Ben-Wu opened the pull request.

wlach commented on analytics_export/analyticsExport.js:
```diff
@@ -1,8 +1,197 @@
+          return value;
+        })
+      )).forEach((value) => {
+        if (!resultsByDate.has(value.date)) {
```
If you were using lodash you could use groupBy here:

https://lodash.com/docs#groupBy

Looking at yarn.lock, it looks like it's being pulled in via some other dependency so probably ok to depend on explicitly.

wlach commented on analytics_export/analyticsExport.js:
```diff
@@ -1,8 +1,197 @@
+      },
+    );
+    const data = await response.json();
+    if (response.ok === false) {
```
I think this less verbose form should still work

```suggestion
    if (!response.ok) {
```

wlach commented on analytics_export/analyticsExport.js:
```diff
@@ -1,8 +1,197 @@
+
+    // Run in for loop to synchronously send each request to avoid hitting rate limit
+    for (const [dimension, measures] of measuresByDimension) {
+      if (!(dimension in dimensionToTableSuffix)) {
```
```suggestion
      if (!dimensionToTableSuffix.includes(dimension)) {
```

Ben-Wu replied:
I think includes is a lodash thing.  Doesn't look like it works here

wlach replied:
hmm, it should be part of the ES API (you know something is lodash because you need to use `_.foo`) but maybe the version of node you're using is too old

wlach commented on analytics_export/analyticsExport.js:
```diff
@@ -1,8 +1,197 @@
+        continue;
+      }
+      for (const measure of measures) {
+        if (!(measure in measureToTablePrefix)) {
```
```suggestion
        if (!measureToTablePrefix.includes(measure)) {
```

wlach commented on analytics_export/bigqueryClient.js:
```diff
@@ -0,0 +1,103 @@
+      { name: 'value', type: 'STRING', mode: 'REQUIRED' },
+    ];
+
+    if (dimension !== null) {
```
```suggestion
    if (dimension) {
```

wlach commented on analytics_export/bigqueryClient.js:
```diff
@@ -0,0 +1,103 @@
+      schema.push({ name: dimensionToTableSuffix[dimension], type: 'STRING', mode: 'REQUIRED' });
+    }
+
+    let tableName;
+    if (dimension !== null) {
+      tableName = `${measureToTablePrefix[measure].name}_by_${dimensionToTableSuffix[dimension]}`;
+    } else {
+      tableName = `${measureToTablePrefix[measure].name}_total`;
+    }
```
Probably can use `const` here?

```suggestion
    const tableName = dimension ? `${measureToTablePrefix[measure].name}_by_${dimensionToTableSuffix[dimension]}` : `${measureToTablePrefix[measure].name}_total`;
```

Ben-Wu commented:
Updated pr with clean up changes including adding prettier.  Also finished unit tests, fixed an issue when trying to load more than 100 days of data into bigquery at the same time, and some minor changes to the output data format.

wlach commented on test/testExport.js:
```diff
@@ -1,13 +1,216 @@
+        );
+      } catch (e) {
+        throwCount += 1;
+      }
```
I'm not entirely clear on why you need three test cases here. Also shouldn't we use jest's built in test for throwing an error? https://jestjs.io/docs/en/expect#tothrowerror

Ben-Wu replied:
I think I was trying to test before, after, and completely out of the boundaries but I mixed up the dates.  And yes should use the builtin things for chai

Ben-Wu replied:
I also noticed that the wrong error is being thrown in the run so that needs to be fixed and checked
